### PR TITLE
Identify dot product pattern (mul followed by a sum) for bf16, and convert it to dot product intrinsics

### PR DESCRIPTION
**Pull request by Kuigesi**

### CODE REVIEW

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+using namespace mlir::triton;
+using namespace mlir::triton::cpu;
+
+namespace {
+
+struct ConvertMulSumToDot
+    : public OpRewritePattern<vector::MultiDimReductionOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::MultiDimReductionOp op,
```
Comment (digantdesai) [inline third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:34]: Is it multi-reduction op even for gemv [1,K] @ [K,N]?

File: third_party/cpu/include/TritonCPUTransforms/Passes.h [hunk 2]
```diff
@@ -29,6 +29,7 @@ createDecomposeFpConversions(bool decomposeBf16Conversions,
                              bool decomposeFp8Conversions);
 std::unique_ptr<OperationPass<ModuleOp>> createOptimizeMasks();
 
+std::unique_ptr<OperationPass<ModuleOp>> createConvertDotProduct();
```
Comment (minjang) [inline third_party/cpu/include/TritonCPUTransforms/Passes.h:32]: nit: add a blank space between line 32 and 33.

File: third_party/cpu/lib/TritonCPUTransforms/CMakeLists.txt [hunk 3]
```diff
@@ -2,6 +2,7 @@ add_triton_library(TritonCPUTransforms
     ConvertUnsupportedOps.cpp
     DecomposeFpConversions.cpp
     OptimizeMasks.cpp
+    ConvertDotProduct.cpp
```
Comment (minjang) [inline third_party/cpu/lib/TritonCPUTransforms/CMakeLists.txt:5]: nit: sort the file names alphabetically.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+} // namespace triton
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::triton;
+using namespace mlir::triton::cpu;
+
+namespace {
+
+// TODO: support SVE and different vector width
```
Comment (minjang) [inline third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:31]: Could you add more comments? This path is currently only for ARM/Neon, not for x86.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+constexpr int vectorBitWidth = 128;
+
+bool isBf16DotProduct(vector::MultiDimReductionOp op, Value &matInput,
```
Comment (minjang) [inline third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:34]: Please add some comments on which patterns can be detected. A simple example would be great.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    auto resTy = cast<VectorType>(op.getType());
+
+    Value matInput;
+    Value vecInput;
+
+    bool isMatch = isBf16DotProduct(op, matInput, vecInput, rewriter);
+    if (!isMatch)
+      return failure();
+
+    auto matInputTy = cast<VectorType>(matInput.getType());
```
Comment (minjang) [inline third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:125]: Similarly, add some comments on how you are converting. Give us a short example.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            loc, resultTypes, intrin, args, LLVM::FastmathFlags::fast);
+        outRes[outIdx] = callIntrOp.getResult(0);
+      }
+    }
+
+    Value res = rewriter.create<arith::ConstantOp>(loc, resTy,
+                                                   rewriter.getZeroAttr(resTy));
+
+    resultTypes = {resTy.getElementType()};
+    intrin = "llvm.aarch64.neon.faddv.f32.v4f32";
```
Comment (minjang) [inline third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:172]: Put some TODO: saying this is hard coded for Neon for now.

**Review by minjang — COMMENTED** [review]
Thanks for this work! It's a very good example how we can leverage a special instruction! So, my comments are really putting some nice (but succinct  and clear) examples and overview of the approach.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    ModuleOp mod = getOperation();
+
+    RewritePatternSet patterns(context);
+
+    patterns.add<ConvertMulSumToDotHorizontalSum>(context);
+
+    if (failed(mlir::applyPatternsAndFoldGreedily(mod, std::move(patterns))))
+      return signalPassFailure();
```
Comment (digantdesai) [inline third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:202]: Any standalone tests you can add for this pass?

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 7 above)
Comment (digantdesai) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:172]: This pass name should have Armv8 + Neon it in TBH

Comment (digantdesai) [inline third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:172]: can LLVM not generate these on its own? Curious if you have tried this and it did something stupid in presence on bfdot intrinsic before

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+    return false;
+
+  const int lanes =
+      vectorBitWidth / lhsTy.getElementType().getIntOrFloatBitWidth();
+  int64_t kVal = lhsTy.getDimSize(1);
+
+  if (outNum < 1)
+    return false;
+
+  if (kVal % lanes != 0)
```
Comment (digantdesai) [inline third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:80]: leave some comments about masking and handling this case

**Review by digantdesai — COMMENTED** [review]
LGTM, leaving to @minjang to decide

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 8 above)
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:202]: Yup, https://github.com/triton-lang/triton-cpu/pull/90, it is already merged!

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 7 above)
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:172]: I need to manually do it to make sure the type of the result checks.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 5 above)
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:34]: https://github.com/triton-lang/triton-cpu/pull/90, it is already merged!

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 7 above)
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:172]: I think Minjang also want to try x86 on it as well, so I did not put the Arm name here. I checked for the platform here:https://github.com/triton-lang/triton-cpu/pull/56/files#diff-ffa39e2a9f7fa74ff0d6bec6641d569d297e1f36aef21a2d446efef9fa224bdeR122

File: third_party/cpu/include/TritonCPUTransforms/OptCommon.h [hunk 10]
```diff
@@ -103,6 +104,22 @@ Value cstLike(Location loc, Value tySrc, T val, PatternRewriter &rewriter) {
...
 
+inline Value shapeCast(Location loc, Value in, VectorType outTy,
+                       PatternRewriter &rewriter) {
+  VectorType inTy = cast<VectorType>(in.getType());
+  assert(outTy.getElementType() == inTy.getElementType());
+  assert(outTy.getNumElements() == inTy.getNumElements());
+  return rewriter.create<vector::ShapeCastOp>(loc, outTy, in);
+}
+
+inline Value shapeCast(Location loc, Value in,
```
[the remark this answers is not in the document]
Comment (Kuigesi) [reply third_party/cpu/include/TritonCPUTransforms/OptCommon.h:115]: I overloaded to simplify the cases where we do not want to manually create the vector type.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    SmallVector<Value> outRes(numOfOuts);
+    SmallVector<Value> mats(numOfOuts);
+
+    Type outResTy = VectorType::get(resLanes, resTy.getElementType());
+
+    Value zeroRes = rewriter.create<arith::ConstantOp>(
+        loc, outResTy, rewriter.getZeroAttr(outResTy));
+    for (int64_t outIdx = 0; outIdx < numOfOuts; outIdx += 1) {
+      outRes[outIdx] = zeroRes;
+      mats[outIdx] = rewriter.create<vector::ExtractOp>(loc, matInput, outIdx);
```
[the remark this answers is not in the document]
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:149]: Yup, matInput is a 3-D vector, I remember directly extract submat from matInput generate non-optimal code. So i put an intermediate array here.

File: third_party/cpu/include/TritonCPUTransforms/Passes.h (hunk 2 above)
Comment (Kuigesi) [reply third_party/cpu/include/TritonCPUTransforms/Passes.h:32]: Added

File: third_party/cpu/lib/TritonCPUTransforms/CMakeLists.txt (hunk 3 above)
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/CMakeLists.txt]: Fixed

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 4 above)
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:31]: Added.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 6 above)
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:164]: Added.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+    auto matInputTy = cast<VectorType>(matInput.getType());
+    auto vecInputTy = cast<VectorType>(vecInput.getType());
+
+    const int lanes =
+        vectorBitWidth / matInputTy.getElementType().getIntOrFloatBitWidth();
+    const int resLanes =
+        vectorBitWidth / resTy.getElementType().getIntOrFloatBitWidth();
+    int64_t kVal = matInputTy.getDimSize(1);
+
+    const int numOfOuts = matInputTy.getDimSize(0);
```
[the remark this answers is not in the document]
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp]: Fixed.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    const int numOfOps = kVal / lanes;
```
[the remark this answers is not in the document]
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp]: Fixed.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+
+    Value zeroRes = rewriter.create<arith::ConstantOp>(
+        loc, outResTy, rewriter.getZeroAttr(outResTy));
+    for (int64_t outIdx = 0; outIdx < numOfOuts; outIdx += 1) {
+      outRes[outIdx] = zeroRes;
+      mats[outIdx] = rewriter.create<vector::ExtractOp>(loc, matInput, outIdx);
+    }
+
+    SmallVector<Type> resultTypes = {outResTy};
+    llvm::StringRef intrin("llvm.aarch64.neon.bfdot.v4f32.v8bf16");
```
[the remark this answers is not in the document]
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp]: Fixed.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+    SmallVector<Value> args;
+
+    for (int64_t idx = 0; idx < numOfOps; idx += 1) {
+      auto subVec = rewriter.create<vector::ExtractOp>(loc, vecInput, idx);
+      for (int64_t outIdx = 0; outIdx < numOfOuts; outIdx += 1) {
+        auto subMat =
+            rewriter.create<vector::ExtractOp>(loc, mats[outIdx], idx);
+        args = {outRes[outIdx], subMat, subVec};
+        auto callIntrOp = rewriter.create<LLVM::CallIntrinsicOp>(
```
[the remark this answers is not in the document]
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:212]: Added.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 9 above)
Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:94]: Added.

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (digantdesai) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp:191]: Please leave a comment here?

File: third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp (hunk 7 above)
Comment (digantdesai) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp]: I meant through some MLIR vector op for doing horizontal sum, but its OK.

Comment (Kuigesi) [reply third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp]: Got it!

**Review by digantdesai — APPROVED** [review]

**Review by digantdesai — APPROVED** [review]

**Review by minjang — APPROVED** [review]
Looks good! Thanks for this exemplary work!

**Pull request merged**